### PR TITLE
Downloading CSV files from remote server

### LINKED ISSUE #15: Downloading files from web

### PR DESCRIPTION
**Pull request by Rumec**
File is now being downloaded and loaded to DB from remote server.

Closes #15

### CODE REVIEW

File: PV260_Project/PresentationLayer/Program.cs [hunk 1]
```diff
@@ -39,9 +39,11 @@ public class Program
                     services.AddTransient<IEmailUi, EmailUi>();
                     services.AddTransient<IDataSetUi, DataSetUi>();
                     services.AddTransient<IDataLoader, CsvFileLoader>();
+                    services.AddTransient<IDataDownloader, CsvFileDownloader>();
                     services.AddTransient<IDiffComputer, DiffComputer>();
                     Console.WriteLine(Directory.GetCurrentDirectory());
-                    services.AddDbContext<StockSystemDbContext>(options => options.UseSqlite("DataSource=..\\..\\..\\..\\DataLayer\\app.db"));
+                    // services.AddDbContext<StockSystemDbContext>(options => options.UseSqlite("DataSource=..\\..\\..\\..\\DataLayer\\app.db"));
+                    services.AddDbContext<StockSystemDbContext>(options => options.UseSqlite("DataSource=/home/rumec/Documents/c#/cv/PV260/PV260_Project/DataLayer/app.db"));
```
Comment (michalSali) [inline PV260_Project/PresentationLayer/Program.cs:46]: I would keep the relative path uncommented, and comment out the absolute/specific path.

**Review by michalSali — APPROVED** [review]

File: PV260_Project/.idea/.idea.PV260_Project/.idea/indexLayout.xml [hunk 2]
```diff
@@ -1,7 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
```
Comment (Bevilton) [inline PV260_Project/.idea/.idea.PV260_Project/.idea/indexLayout.xml:1]: I would add the `.idea` folder to .gitignore

File: PV260_Project/BusinessLayer/DataLoading/CsvFileDownloader.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            {
+                throw new WebException($"Responded with {responseMessage.StatusCode}");
+            }
+        }
+        catch (Exception e)
+        {
+            throw new DataLoaderException($"Error while downloading csv file from '{path}'", e);
+        }
+
+        var config = new CsvConfiguration(CultureInfo.InvariantCulture)
```
Comment (Bevilton) [inline PV260_Project/BusinessLayer/DataLoading/CsvFileDownloader.cs:36]: These lines are same as in the `CsvFileLoader`, sounds like its rework time :D Also those two interfaces are same, to I would use just one interface and both loaders can implement same interface. The shared logic can be put in some abstract class which can be inherited by the loaders. The logic inside `CsvOperation` can also be put to this abstract class.

File: PV260_Project/BusinessLayer/DataLoading/Utils/CsvOperations.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+            holding.MarketValue = value;
+            holding.Weight = double.Parse(csvLine.Weight.Remove(csvLine.Weight.Length - 1),
+                CultureInfo.InvariantCulture);
+            file.Holdings.Add(holding);
+        }
+
+        return file;
+    }
+
+    public static DataSet ParseCsvData(string path, CsvReader csv)
```
Comment (Bevilton) [inline PV260_Project/BusinessLayer/DataLoading/Utils/CsvOperations.cs:36]: When `path` can be both path to file and url, I would change its name to something more meaningful :D

File: PV260_Project/PresentationLayer/UI/DataSetUi.cs [hunk 5]
```diff
@@ -18,84 +12,135 @@ public class DataSetUi : BaseUi, IDataSetUi
...
         }
 
-        private async Task DownloadCurrentFile() {
+        private async Task DownloadCurrentFile()
+        {
+            var input = "";
 
+            while (input! != "b")
+            {
+                Console.WriteLine("Path: ('b' for back)");
```
Comment (Bevilton) [inline PV260_Project/PresentationLayer/UI/DataSetUi.cs:76]: I see you got inspired by the `LoadFile` method :D but I'm not sure, if downloading file needs path. I think that the path should be written somewhere in code or in config so users won't need to remember the url. Also this method looks a lot like the `LoadFile`. Maybe refinement? :D If you are bored... :D

**Comment by Bevilton** [comment]
Gj

File: PV260_Project/PresentationLayer/Program.cs (hunk 1 above)
Comment (Bevilton) [reply PV260_Project/PresentationLayer/Program.cs:46]: Try if this works (or something similar) on your OS. `Path.Combine(Directory.GetParent(Environment.CurrentDirectory).Parent.Parent.FullName, "..", "DataLayer", "app.db")`

File: PV260_Project/PresentationLayer/UI/DataSetUi.cs (hunk 5 above)
Comment (Rumec) [reply PV260_Project/PresentationLayer/UI/DataSetUi.cs:76]: I did not know how to implement it actually. I think this actually makes sense for manual downloading the files (cuz user can choose his own destination) and for automatic downloads, there would be a URL in DB.

File: PV260_Project/BusinessLayer/DataLoading/Utils/CsvOperations.cs (hunk 4 above)
Comment (Rumec) [reply PV260_Project/BusinessLayer/DataLoading/Utils/CsvOperations.cs:36]: makes sense!

File: PV260_Project/BusinessLayer/DataLoading/CsvFileDownloader.cs (hunk 3 above)
Comment (Rumec) [reply PV260_Project/BusinessLayer/DataLoading/CsvFileDownloader.cs:36]: We were told at the lectures not to use abstract classes :D. 

Those are not 1:1 the same - in this file, I am using a different type of reader (not reading from a file but from a string). 

However, I am no C# master :D so it is possible to unify somehow those 2 approaches (stringreader and streamreader) I am all in!

File: PV260_Project/PresentationLayer/Program.cs (hunk 1 above)
Comment (Rumec) [reply PV260_Project/PresentationLayer/Program.cs:46]: Sorry, I forgot to change it in this commit.

The `Path.Combine` approach did not work for some reason. I will add a NOTE informing UNIX system users about this. AIFAK Mr. Goth uses Windows, so I will keep it like this.

File: PV260_Project/BusinessLayer/DataLoading/CsvFileDownloader.cs (hunk 3 above)
Comment (Bevilton) [reply PV260_Project/BusinessLayer/DataLoading/CsvFileDownloader.cs:36]: Wtf who told us about not using abstract classes? :D They are very useful. It is different type of reader but the interface looks the same

File: PV260_Project/PresentationLayer/UI/DataSetUi.cs (hunk 5 above)
Comment (Bevilton) [reply PV260_Project/PresentationLayer/UI/DataSetUi.cs:76]: Ok, good point

**Review by Bevilton — APPROVED** [review]

**Review by MatejMichalek — APPROVED** [review]

**Pull request merged**